log/levels/levels.go

```diff
@@ -29,8 +29,54 @@ type Levels struct {
 	warnValue  string
 	errorValue string
 	critValue  string
+
+	// FilterLogLevel is an option allowing logs below a certain level to be
+	// discarded rather than logged. For example, a consumer may set LogLevel
+	// to levels.Warn, and any logs to Debug or Info would become no-ops.
+	filterLogLevel LogLevel
+
+	// Promote errors
+	promoteErrors       bool
+	errorKey            string
+	promoteErrorToLevel LogLevel
+}
+
+// levelCommittedLogger embeds a log level that the user has selected and prepends
+// the associated level string as a prefix to every log line. Log level may be
+// affected by error promotion; and log lines may be suppressed based on the value
+// of FilterLogLevel.
+type levelCommittedLogger struct {
```

Comment:
For a first pass, log.Context.With() also creates a new struct, so the `Debug()` call should have the same or fewer allocations. The `Log()` call in `levelCommittedLogger` might have a few more allocations since it does a `log.Context.With()` and a `log.Context.WithPrefix()`, but they will immediately go out of scope and become eligible for GC.

I don't envision it being a huge difference, but I will add some memory benchmarks to the test suite to check!